sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala

```diff
@@ -78,9 +78,42 @@ case class ProjectExec(projectList: Seq[NamedExpression], child: SparkPlan)
     }
   }
 
-  override def outputOrdering: Seq[SortOrder] = child.outputOrdering
```

Comment:
My original thought is to support the cases where `requiresChildDistribution` refers nested fields created and aliased in `Project`, as the example I showed.

The example you give is another kind of case. In order to support, if we wannt, at least we need to improve expression canonicalization and partitioning/distribution matching.